### PR TITLE
Enhancement/1824 error message

### PR DESCRIPTION
**Pull request by eugene-manuilov**
## Summary

Addresses issue #1824 

## Relevant technical choices

* Renamed `is-wp-error.js` file to be just `errors.js` and moved `isPermissionScopeError` function to the new file sinse it is a better place for such kind of functions.
* User information:
  * Didn't use `list_users` capabilities because it's not needed since we can't rely on the standard REST API users endpoint 
because it can be disabled or modified by site developers.
  * Added owner display name property to module information to be able to properly compose error message.
  * Modified the Search Console module class to implement Module_With_Owner interface and use global `Owner_ID` as module owner.

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: includes/Core/Modules/Modules.php [hunk 1]
```diff
@@ -816,6 +825,15 @@ private function prepare_module_data_for_response( Module $module ) {
 			'dependencies' => $this->get_module_dependencies( $module->slug ),
 			'dependants'   => $this->get_module_dependants( $module->slug ),
 		);
+
+		if ( $module instanceof Module_With_Owner ) {
+			$owner_id = $module->get_owner_id();
+			if ( $owner_id ) {
+				$module_data['owner'] = get_the_author_meta( 'display_name', $owner_id );
```
Comment (felixarntz) [inline includes/Core/Modules/Modules.php:832]: Per the ACs, this should use the username (kind of the "user slug"), not the display name. Furthermore, I think it would make more sense to alter the data format here:

* The REST API endpoints should be future proof. Exposing a name as something like `owner` can quickly get us trapped in a place where we don't have enough flexibility to change the format later - for example, it's not clear whether `owner` refers to an ID, slug, display name, ... I think we should make `owner` an object (associative array) here and expose `id` and `login` on it for now.
* API response data should be consistent, there shouldn't be any properties that exist on some responses but are missing somewhere else. As such, if there is no owner, or if the module is not a `Module_With_Owner`, this should still be included, as `null`.

File: includes/Core/Modules/Modules.php [hunk 2]
```diff
@@ -882,6 +900,11 @@ private function get_module_schema() {
 					),
 					'readonly'    => true,
 				),
+				'owner'        => array(
+					'type'        => 'string',
+					'description' => __( 'Owner of the module.', 'google-site-kit' ),
+					'readonly'    => true,
+				),
```
Comment (felixarntz) [inline includes/Core/Modules/Modules.php:907]: See above, let's make this an object with `id` and `login`.

File: includes/Core/Modules/Modules.php [hunk 3]
```diff
@@ -882,6 +900,11 @@ private function get_module_schema() {
 					),
 					'readonly'    => true,
 				),
+				'owner'        => array(
```
Comment (felixarntz) [inline includes/Core/Modules/Modules.php:903]: Maybe we could bake the `list_users` permission check into here instead of JS (see my other comment above)? That would actually simplify things a bit, plus it would ensure we don't even surface this data to users that shouldn't see it.

Since that is based on user permissions, I'd argue here we could alter even the base response: The `owner` property would only be included (either as object or `null`) if the current user can `list_users`.

File: includes/Modules/Search_Console.php [hunk 4]
```diff
@@ -589,4 +591,17 @@ protected function setup_assets() {
...
+	 * Gets an owner ID for the module.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return int Owner ID.
+	 */
+	public function get_owner_id() {
+		$owner = new Owner_ID( $this->options );
+		return $owner->get();
+	}
```
Comment (felixarntz) [inline includes/Modules/Search_Console.php:605]: 👍  Since Search Console doesn't have a specific owner, makes sense to rely on the global "owner" here.

**Review by felixarntz — CHANGES_REQUESTED** [review]
This is on the right track, but there are a few things to revisit. I think we can simplify quite a bit.

File: assets/js/util/insufficient-permissions-error-description.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	if ( _x( 'Analytics', 'Service name', 'google-site-kit' ) === moduleName ) {
+		if ( error.match( /account/i ) ) {
+			message = __( `Your Google account does not have sufficient permissions for this Analytics account, so you won't be able to see stats from it on the Site Kit dashboard.`, 'google-site-kit' );
+		} else if ( error.match( /property/i ) ) {
+			message = __( `Your Google account does not have sufficient permissions for this Analytics property, so you won't be able to see stats from it on the Site Kit dashboard.`, 'google-site-kit' );
+		} else if ( error.match( /view/i ) ) {
+			message = __( `Your Google account does not have sufficient permissions for this Analytics view, so you won't be able to see stats from it on the Site Kit dashboard.`, 'google-site-kit' );
+		}
+	} else if ( _x( 'Search Console', 'Service name', 'google-site-kit' ) === moduleName ) {
+		message = __( `Your Google account does not have sufficient permissions for this Search Console property, so you won't be able to see stats from it on the Site Kit dashboard.`, 'google-site-kit' );
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/util/insufficient-permissions-error-description.js:47]: Using a placeholder makes the translatable string harder to translate. These four messages are good as is because they are easy to understand and translate. Plus, there is no need for a placeholder because module names for these messages are constant and won't change.

File: assets/js/components/StoreErrorNotice.js [hunk 6]
```diff
@@ -25,22 +25,38 @@ import PropTypes from 'prop-types';
...
 import Data from 'googlesitekit-data';
-import { isPermissionScopeError } from '../googlesitekit/datastore/user/utils/is-permission-scope-error';
+import { STORE_NAME as CORE_MODULES } from '../googlesitekit/modules/datastore/constants';
+import { isPermissionScopeError, isInsufficientPermissionsError } from '../util/errors';
+import { getInsufficientPermissionsErrorDescription } from '../util/insufficient-permissions-error-description';
 import ErrorText from '../components/error-text';
 const { useSelect } = Data;
 
-export default function StoreErrorNotice( { storeName, shouldDisplayError = () => true } ) {
+function StoreErrorNotice( { moduleSlug, moduleName, storeName, shouldDisplayError } ) {
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/components/StoreErrorNotice.js]: Yep, updated.

File: assets/js/components/StoreErrorNotice.js [hunk 7]
```diff
@@ -25,22 +25,38 @@ import PropTypes from 'prop-types';
...
 	const error = useSelect( ( select ) => select( storeName ).getError() );
+	const module = useSelect( ( select ) => select( CORE_MODULES ).getModule( moduleSlug ) || {} );
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/components/StoreErrorNotice.js]: Yeah, removed the default value and passed module as is to the getInsufficientPermissionsErrorDescription function that handles empty module state.

File: assets/js/components/StoreErrorNotice.js [hunk 8]
```diff
@@ -25,22 +25,38 @@ import PropTypes from 'prop-types';
...
 	// Do not display if no error, or if the error is for missing scopes.
 	if ( ! error || isPermissionScopeError( error ) || ! shouldDisplayError( error ) ) {
 		return null;
 	}
 
-	return <ErrorText message={ error.message } reconnectURL={ error.data?.reconnectURL } />;
+	let message = error.message;
+	if ( isInsufficientPermissionsError( error ) ) {
+		message = getInsufficientPermissionsErrorDescription( message, moduleName, module?.owner );
+	}
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/components/StoreErrorNotice.js]: Updated.

File: assets/js/components/StoreErrorNotice.js [hunk 9]
```diff
@@ -25,22 +25,38 @@ import PropTypes from 'prop-types';
...
+
+	return <ErrorText message={ message } reconnectURL={ error.data?.reconnectURL } />;
 }
 
 StoreErrorNotice.propTypes = {
+	module: PropTypes.string.isRequired,
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/components/StoreErrorNotice.js]: :man_facepalming: yep, updated.

File: assets/js/components/StoreErrorNotice.js [hunk 10]
```diff
@@ -25,22 +25,38 @@ import PropTypes from 'prop-types';
...
+	moduleName: PropTypes.string.isRequired,
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/components/StoreErrorNotice.js]: Removed.

File: assets/js/modules/adsense/components/common/ErrorNotice.js [hunk 11]
```diff
@@ -30,5 +31,13 @@ import StoreErrorNotice from '../.././../../components/StoreErrorNotice';
 
 export default function ErrorNotice() {
 	const shouldDisplayError = useCallback( ( error ) => undefined === errorToStatus( error ), [] );
-	return <StoreErrorNotice storeName={ STORE_NAME } shouldDisplayError={ shouldDisplayError } />;
+
+	return (
+		<StoreErrorNotice
+			module="adsense"
+			moduleName={ _x( 'AdSense', 'Service name', 'google-site-kit' ) }
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/modules/adsense/components/common/ErrorNotice.js]: Updated.

File: assets/js/modules/adsense/components/common/ErrorNotice.js [hunk 12]
```diff
@@ -30,5 +31,13 @@ import StoreErrorNotice from '../.././../../components/StoreErrorNotice';
 
 export default function ErrorNotice() {
 	const shouldDisplayError = useCallback( ( error ) => undefined === errorToStatus( error ), [] );
-	return <StoreErrorNotice storeName={ STORE_NAME } shouldDisplayError={ shouldDisplayError } />;
+
+	return (
+		<StoreErrorNotice
+			module="adsense"
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/modules/adsense/components/common/ErrorNotice.js]: Updated.

File: assets/js/modules/analytics/components/common/AccountCreate/index.js [hunk 13]
```diff
@@ -134,7 +135,12 @@ export default function AccountCreate() {
 
 	return (
 		<div>
-			<StoreErrorNotice storeName={ STORE_NAME } />
+			<StoreErrorNotice
+				module="analytics"
+				moduleName={ _x( 'Analytics', 'Service name', 'google-site-kit' ) }
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/common/AccountCreate/index.js]: Updated.

File: assets/js/modules/analytics/components/common/AccountCreateLegacy.js [hunk 14]
```diff
@@ -58,7 +58,12 @@ export default function AccountCreateLegacy() {
 
 	return (
 		<div>
-			<StoreErrorNotice storeName={ STORE_NAME } />
+			<StoreErrorNotice
+				module="analytics"
+				moduleName={ _x( 'Analytics', 'Service name', 'google-site-kit' ) }
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/common/AccountCreateLegacy.js]: Updated.

File: assets/js/modules/analytics/components/settings/SettingsForm.js [hunk 15]
```diff
@@ -35,7 +40,12 @@ import { STORE_NAME } from '../../datastore/constants';
 export default function SettingsForm() {
 	return (
 		<div className="googlesitekit-analytics-settings-fields">
-			<StoreErrorNotice storeName={ STORE_NAME } />
+			<StoreErrorNotice
+				module="analytics"
+				moduleName={ _x( 'Analytics', 'Service name', 'google-site-kit' ) }
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/settings/SettingsForm.js]: Updated.

File: assets/js/modules/analytics/components/settings/SettingsView.js [hunk 16]
```diff
@@ -54,7 +54,7 @@ export default function SettingsView() {
 		<div className="googlesitekit-setup-module googlesitekit-setup-module--analytics">
 
 			{ /* Prevent showing ExistingTagError and general error notice at the same time. */ }
-			{ ( ! hasExistingTag || hasExistingTagPermission ) && <StoreErrorNotice storeName={ STORE_NAME } /> }
+			{ ( ! hasExistingTag || hasExistingTagPermission ) && <StoreErrorNotice module="analytics" moduleName={ _x( 'Analytics', 'Service name', 'google-site-kit' ) } storeName={ STORE_NAME } /> }
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/settings/SettingsView.js]: Updated.

File: assets/js/modules/analytics/components/setup/SetupForm.js [hunk 17]
```diff
@@ -80,11 +80,13 @@ export default function SetupForm( { finishSetup } ) {
...
 	return (
-		<form
-			className="googlesitekit-analytics-setup__form"
-			onSubmit={ submitForm }
-		>
-			<StoreErrorNotice storeName={ STORE_NAME } />
+		<form className="googlesitekit-analytics-setup__form" onSubmit={ submitForm }>
+			<StoreErrorNotice
+				module="analytics"
+				moduleName={ _x( 'Analytics', 'Service name', 'google-site-kit' ) }
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/setup/SetupForm.js]: Updated.

File: assets/js/modules/optimize/components/settings/SettingsForm.js [hunk 18]
```diff
@@ -42,7 +42,12 @@ export default function SettingsForm() {
 
 	return (
 		<div className="googlesitekit-optimize-settings-fields">
-			<StoreErrorNotice storeName={ STORE_NAME } />
+			<StoreErrorNotice
+				module="optimize"
+				moduleName={ _x( 'Optimize', 'Service name', 'google-site-kit' ) }
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/modules/optimize/components/settings/SettingsForm.js]: Updated.

File: assets/js/util/insufficient-permissions-error-description.js [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+ * Gets a description for an insufficient permissions error.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string} error Original error message.
+ * @param {string} moduleName Module name.
+ * @param {string} ownerName Module owner name.
+ * @return {string} Error description.
+ */
+export function getInsufficientPermissionsErrorDescription( error, moduleName, ownerName ) {
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/util/insufficient-permissions-error-description.js]: Yep, updated.

File: assets/js/util/insufficient-permissions-error-description.js [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+	let message = '';
+	let userInfo = '';
+
+	if ( _x( 'Analytics', 'Service name', 'google-site-kit' ) === moduleName ) {
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/util/insufficient-permissions-error-description.js]: Yes, agree. Updated.

File: assets/js/util/insufficient-permissions-error-description.js [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+			message = __( `Your Google account does not have sufficient permissions for this Analytics view, so you won't be able to see stats from it on the Site Kit dashboard.`, 'google-site-kit' );
+		}
+	} else if ( _x( 'Search Console', 'Service name', 'google-site-kit' ) === moduleName ) {
+		message = __( `Your Google account does not have sufficient permissions for this Search Console property, so you won't be able to see stats from it on the Site Kit dashboard.`, 'google-site-kit' );
+	}
+
+	if ( ! message ) {
+		message = sprintf(
+			/* translators: %s: module name */
+			__( `Your Google account does not have sufficient permissions for this %s data, so you won't be able to see stats from it on the Site Kit dashboard.`, 'google-site-kit' ),
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/util/insufficient-permissions-error-description.js]: Updated.

File: assets/js/util/insufficient-permissions-error-description.js [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+			moduleName,
+		);
+	}
+
+	if ( ownerName ) {
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/util/insufficient-permissions-error-description.js]: Added `list_users` check to the Modules class.

File: includes/Core/Modules/Modules.php (hunk 1 above)
Comment (eugene-manuilov) [reply includes/Core/Modules/Modules.php]: Good point. Updated.

File: includes/Core/Modules/Modules.php (hunk 2 above)
Comment (eugene-manuilov) [reply includes/Core/Modules/Modules.php]: Yep, updated.

File: includes/Core/Modules/Modules.php (hunk 3 above)
Comment (eugene-manuilov) [reply includes/Core/Modules/Modules.php:911]: Yes, a good idea. Added capability check to the Modules class. Thanks!

File: assets/js/components/notifications/data-error.js [hunk 23]
```diff
@@ -46,23 +45,29 @@ import InvalidCredentialsWarning from './invalid-credentials-warning';
  *
  * @return {WPElement} CTA component with data error message.
  */
-const getDataErrorComponent = ( moduleName, error, inGrid = false, fullWidth = false, createGrid = false, errorObj = {} ) => {
-	const reconnectUrl = errorObj?.data?.reconnectURL;
-	const description = reconnectUrl ? <ErrorText message={ error } reconnectURL={ reconnectUrl } /> : error;
+function getDataErrorComponent( moduleSlug, moduleName, error, inGrid = false, fullWidth = false, createGrid = false, errorObj = {} ) {
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/components/notifications/data-error.js]: Updated.

File: assets/js/components/higherorder/withdata.js [hunk 24]
```diff
@@ -265,7 +265,7 @@ const withData = (
 
 			// If we have an error, display the DataErrorComponent.
 			if ( error ) {
-				return ( 'string' !== typeof error ) ? error : getDataErrorComponent( moduleName, error, layoutOptions.inGrid, layoutOptions.fullWidth, layoutOptions.createGrid, data );
+				return ( 'string' !== typeof error ) ? error : getDataErrorComponent( module, moduleName, error, layoutOptions.inGrid, layoutOptions.fullWidth, layoutOptions.createGrid, data );
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/components/higherorder/withdata.js]: Updated.

File: assets/js/modules/adsense/components/dashboard/AdSenseDashboardWidget.js [hunk 25]
```diff
@@ -174,7 +174,7 @@ class AdSenseDashboardWidget extends Component {
 								</div>
 							}
 							{ ! receivingData && (
-								error ? getDataErrorComponent( _x( 'AdSense', 'Service name', 'google-site-kit' ), error, true, true, true, errorObj ) : getNoDataComponent( _x( 'AdSense', 'Service name', 'google-site-kit' ), true, true, true )
+								error ? getDataErrorComponent( 'adsense', _x( 'AdSense', 'Service name', 'google-site-kit' ), error, true, true, true, errorObj ) : getNoDataComponent( _x( 'AdSense', 'Service name', 'google-site-kit' ), true, true, true )
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/modules/adsense/components/dashboard/AdSenseDashboardWidget.js]: Updated.

File: assets/js/modules/analytics/components/dashboard/AnalyticsDashboardWidget.js [hunk 26]
```diff
@@ -183,7 +183,7 @@ class AnalyticsDashboardWidget extends Component {
 							</div>
 							{ /* Data issue: on error display a notification. On missing data: display a CTA. */ }
 							{ ! receivingData && (
-								error ? getDataErrorComponent( _x( 'Analytics', 'Service name', 'google-site-kit' ), error, true, true, true, errorObj ) : getNoDataComponent( _x( 'Analytics', 'Service name', 'google-site-kit' ), true, true, true )
+								error ? getDataErrorComponent( 'analytics', _x( 'Analytics', 'Service name', 'google-site-kit' ), error, true, true, true, errorObj ) : getNoDataComponent( _x( 'Analytics', 'Service name', 'google-site-kit' ), true, true, true )
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/dashboard/AnalyticsDashboardWidget.js]: Updated.

File: assets/js/modules/analytics/components/dashboard/DashboardAllTrafficWidget.js [hunk 27]
```diff
@@ -92,7 +92,7 @@ function DashboardAllTrafficWidget() {
 	} );
 
 	if ( ! loading && error ) {
-		return getDataErrorComponent( _x( 'Analytics', 'Service name', 'google-site-kit' ), error.message, true, true, false, error );
+		return getDataErrorComponent( 'analytics', _x( 'Analytics', 'Service name', 'google-site-kit' ), error.message, true, true, false, error );
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/dashboard/DashboardAllTrafficWidget.js]: Updated.

File: assets/js/modules/analytics/components/dashboard/LegacyAnalyticsAllTraffic.js [hunk 28]
```diff
@@ -92,7 +92,7 @@ class LegacyAnalyticsAllTraffic extends Component {
 		}
 
 		if ( ! receivingData ) {
-			return error ? getDataErrorComponent( _x( 'Analytics', 'Service name', 'google-site-kit' ), error, true, true, true, errorObj ) : getNoDataComponent( _x( 'Analytics', 'Service name', 'google-site-kit' ), true, true, true );
+			return error ? getDataErrorComponent( 'analytics', _x( 'Analytics', 'Service name', 'google-site-kit' ), error, true, true, true, errorObj ) : getNoDataComponent( _x( 'Analytics', 'Service name', 'google-site-kit' ), true, true, true );
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/dashboard/LegacyAnalyticsAllTraffic.js]: Updated.

File: assets/js/components/notifications/data-error.js [hunk 29]
```diff
@@ -46,23 +45,29 @@ import InvalidCredentialsWarning from './invalid-credentials-warning';
...
+		message = getInsufficientPermissionsErrorDescription( message, moduleName, modulesData[ moduleSlug ]?.owner );
+	}
+
+	const reconnectUrl = errorObj?.data?.reconnectURL;
+	const description = reconnectUrl ? <ErrorText message={ message } reconnectURL={ reconnectUrl } /> : message;
+	const cta = <CTA title={ title } description={ description } error />;
 
 	// This is to handle token expired error specifically.
-	if ( 'Invalid Credentials' === error ) {
+	if ( 'Invalid Credentials' === message ) {
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/components/notifications/data-error.js:64]: Slightly reworked `getDataErrorComponent` function and now `error` is an error object itself. If you want, I can update it to be `errorMessage` instead. Would it work?

[the remark this answers is not in the document]
Comment (felixarntz) [reply assets/js/components/notifications/data-error.js:64]: We shouldn't modify the signature here. I agree the current one is completely weird, but there is a lot of complexity going on here that I'd rather not deal with as part of this issue. The last parameter here is `errorObj` anyway, so let's rely on that. `error` (which maybe should indeed be renamed `errorMessage` for clarity) may be different, depending on what `getDataError` does with the returned API response.

Hence, let's keep the signature of this function as is, rely on `error` (or `errorMessage` if we rename the variable) in this line, and on `errorObj` (if provided) for the other logic which is the main part of this PR.

File: assets/js/components/higherorder/withdata.js [hunk 30]
```diff
@@ -265,7 +267,7 @@ const withData = (
 
 			// If we have an error, display the DataErrorComponent.
 			if ( error ) {
-				return ( 'string' !== typeof error ) ? error : getDataErrorComponent( moduleName, error, layoutOptions.inGrid, layoutOptions.fullWidth, layoutOptions.createGrid, data );
+				return ( 'string' !== typeof error ) ? error : getDataErrorComponent( module, errorObj, layoutOptions.inGrid, layoutOptions.fullWidth, layoutOptions.createGrid );
```
Comment (felixarntz) [inline assets/js/components/higherorder/withdata.js:270]: Let's revert the changes in this file accordingly.

File: assets/js/modules/adsense/components/dashboard/AdSenseDashboardWidget.js [hunk 31]
```diff
@@ -174,7 +174,7 @@ class AdSenseDashboardWidget extends Component {
 								</div>
 							}
 							{ ! receivingData && (
-								error ? getDataErrorComponent( _x( 'AdSense', 'Service name', 'google-site-kit' ), error, true, true, true, errorObj ) : getNoDataComponent( _x( 'AdSense', 'Service name', 'google-site-kit' ), true, true, true )
+								error ? getDataErrorComponent( 'adsense', errorObj, true, true, true ) : getNoDataComponent( _x( 'AdSense', 'Service name', 'google-site-kit' ), true, true, true )
```
Comment (felixarntz) [inline assets/js/modules/adsense/components/dashboard/AdSenseDashboardWidget.js:177]: Let's update all the usages of `getDataErrorComponent` to still pass the error message as second parameter, and `errorObj` as last parameter where it currently is passed. The only thing that should be changed is passing the module slug instead of the module name.

File: assets/js/components/notifications/data-error.js [hunk 32]
```diff
@@ -37,32 +35,37 @@ import InvalidCredentialsWarning from './invalid-credentials-warning';
...
- * @param {Object}  errorObj   Error related data.
- *
  * @return {WPElement} CTA component with data error message.
  */
-const getDataErrorComponent = ( moduleName, error, inGrid = false, fullWidth = false, createGrid = false, errorObj = {} ) => {
-	const reconnectUrl = errorObj?.data?.reconnectURL;
-	const description = reconnectUrl ? <ErrorText message={ error } reconnectURL={ reconnectUrl } /> : error;
+function getDataErrorComponent( moduleSlug, error, inGrid = false, fullWidth = false, createGrid = false ) {
+	const modulesData = getModulesData();
+	const module = modulesData[ moduleSlug ];
```
Comment (felixarntz) [inline assets/js/components/notifications/data-error.js:47]: This works, but I think we should be able to modify it to rely on `useSelect` and `core/modules` selector `getModule`? Just a minor modernization and one less usage of `getModulesData`. That would also ensure the objects received by `getInsufficientPermissionsErrorDescriptions` are of the same shape for both the legacy and the new code.

File: assets/js/modules/search-console/components/dashboard/DashboardClicksWidget.js [hunk 33]
```diff
@@ -67,7 +67,7 @@ function DashboardClicksWidget() {
 
 	if ( error ) {
 		trackEvent( 'plugin_setup', 'search_console_error', error.message );
-		return getDataErrorComponent( __( 'Search Console', 'google-site-kit' ), error.message );
+		return getDataErrorComponent( 'search-console', error );
```
Comment (felixarntz) [inline assets/js/modules/search-console/components/dashboard/DashboardClicksWidget.js:70]: Kind of unrelated to this PR, but let's update this to `getDataErrorComponent( 'search-console', error.message, false, false, false, error )` so that everything is passed as expected.

File: assets/js/modules/search-console/components/dashboard/DashboardImpressionsWidget.js [hunk 34]
```diff
@@ -67,7 +67,7 @@ function DashboardImpressionsWidget() {
 
 	if ( error ) {
 		trackEvent( 'plugin_setup', 'search_console_error', error.message );
-		return getDataErrorComponent( __( 'Search Console', 'google-site-kit' ), error.message );
+		return getDataErrorComponent( 'search-console', error );
```
Comment (felixarntz) [inline assets/js/modules/search-console/components/dashboard/DashboardImpressionsWidget.js:70]: See above, let's update this to `getDataErrorComponent( 'search-console', error.message, false, false, false, error )` so that everything is passed as expected.

File: assets/js/util/insufficient-permissions-error-description.js [hunk 35]
```diff
@@ -0,0 +1,76 @@
...
+ *
+ * @param {string} error Original error message.
+ * @param {Object} module Module data.
+ * @return {string} Error description.
+ */
+export function getInsufficientPermissionsErrorDescription( error, module ) {
+	const {
+		slug = '',
+		name = '',
+		owner = '',
```
Comment (felixarntz) [inline assets/js/util/insufficient-permissions-error-description.js:37]: This should probably use `null` as default to indicate it's expected to be an object.

**Review by felixarntz — CHANGES_REQUESTED** [review]
This is looking great now overall, but let's revert the change of removing `error` as parameter of `getDataErrorComponent` because it may be different from the error object's message, as it relies on the `getDataError` callback function. This way we keep potential legacy regressions at a minimum. The current signature of that legacy function is super weird, but I think it's better to keep it that way for now.

File: assets/js/components/notifications/data-error.js (hunk 29 above)
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/components/notifications/data-error.js]: Ok, got it. Reverted.

File: assets/js/components/higherorder/withdata.js (hunk 30 above)
Comment (eugene-manuilov) [reply assets/js/components/higherorder/withdata.js]: Yep, updated.

File: assets/js/modules/adsense/components/dashboard/AdSenseDashboardWidget.js (hunk 31 above)
Comment (eugene-manuilov) [reply assets/js/modules/adsense/components/dashboard/AdSenseDashboardWidget.js]: Yep, updated.

File: assets/js/modules/search-console/components/dashboard/DashboardClicksWidget.js (hunk 33 above)
Comment (eugene-manuilov) [reply assets/js/modules/search-console/components/dashboard/DashboardClicksWidget.js]: Updated.

File: assets/js/modules/search-console/components/dashboard/DashboardImpressionsWidget.js (hunk 34 above)
Comment (eugene-manuilov) [reply assets/js/modules/search-console/components/dashboard/DashboardImpressionsWidget.js]: Updated.

File: assets/js/util/insufficient-permissions-error-description.js (hunk 35 above)
Comment (eugene-manuilov) [reply assets/js/util/insufficient-permissions-error-description.js]: Yep, updated.

File: assets/js/components/notifications/data-error.js (hunk 32 above)
Comment (eugene-manuilov) [reply assets/js/components/notifications/data-error.js:48]: @felixarntz I am not sure that it will work correctly because we use `getDataErrorComponent` as a regular function. If we used it as a component, we would be able to use hooks there, right?

**Review by felixarntz — APPROVED** [review]
Awesome!
